### PR TITLE
Add thresholding image filter and helper function

### PR DESCRIPTION
**Pull request by russelldj**
Use VXL to perform either absolute or approximate-percentile thresholding. Add general templated functions in a header file.

### CODE REVIEW

**Comment by russelldj** [comment]
This also removes definitions from `convert_image.cxx` which are now duplicated in this helper file.

File: arrows/vxl/convert_image.cxx [hunk 1]
```diff
@@ -3,6 +3,7 @@
 // https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
 
 #include "convert_image.h"
+#include "image_statistics.h"
 
 #include <arrows/vxl/image_container.h>
 #include <vital/range/iota.h>
```
Comment (mwoehlke-kitware) [inline arrows/vxl/convert_image.cxx:9]: BTW, no longer needed?

File: arrows/vxl/CMakeLists.txt [hunk 2]
```diff
@@ -22,10 +22,12 @@ set(vxl_headers_public
   high_pass_filter.h
   image_container.h
   image_io.h
+  image_statistics.h
```
Comment (mwoehlke-kitware) [inline arrows/vxl/CMakeLists.txt:25]: Is this really (meant to be) a public header?

File: arrows/vxl/image_statistics.h [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+// ----------------------------------------------------------------------------
+// Resultant image is true where pixel is greater than a percentile threshold
+template < typename PixType >
+void
+percentile_threshold_above(
+  const vil_image_view< PixType >& src, const double percentile,
+  vil_image_view< bool >& dst, unsigned sampling_points = 1000 )
+{
+  percentile_threshold_above(
+    src, std::vector< double >( 1, percentile ), dst, sampling_points );
```
Comment (mwoehlke-kitware) [inline arrows/vxl/image_statistics.h:139]:
```suggestion
    src, std::vector< double >{ 1, percentile }, dst, sampling_points );
```

File: arrows/vxl/threshold.cxx [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+// OSI-approved BSD 3-Clause License. See top-level LICENSE file or
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+#include "threshold.h"
+
+#include "image_statistics.h"
+
+#include <arrows/vxl/image_container.h>
+#include <vital/range/iota.h>
+#include <vital/util/enum_converter.h>
```
Comment (mwoehlke-kitware) [inline arrows/vxl/threshold.cxx:11]:
```suggestion
#include <arrows/vxl/image_container.h>

#include <vital/util/enum_converter.h>

#include <vital/range/iota.h>
```

(Note: `vital/range` is considered "less specific" than the rest of KWIVER.)

File: arrows/vxl/threshold.cxx [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+#include <vil/vil_convert.h>
+#include <vil/vil_image_view.h>
+#include <vil/vil_math.h>
+#include <vil/vil_plane.h>
+
+#include <cstdlib>
+#include <limits>
+#include <random>
+#include <type_traits>
```
Comment (mwoehlke-kitware) [inline arrows/vxl/threshold.cxx:21]:
```suggestion
#include <limits>
#include <random>
#include <type_traits>

#include <cstdlib>
```

File: arrows/vxl/threshold.cxx [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+ENUM_CONVERTER( mode_converter, threshold_mode, { "absolute", MODE_absolute },
+                { "percentile", MODE_percentile } );
+
+// ----------------------------------------------------------------------------
+// Private implementation class
+class threshold::priv
+{
+public:
+  priv() = default;
+
```
Comment (mwoehlke-kitware) [inline arrows/vxl/threshold.cxx:44]: This doesn't appear useful?
```suggestion
```

File: arrows/vxl/threshold.h [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+/// Threshold an image using different schemes.
+///
+/// Use either an absolute threshold or one based on percentiles.
+class KWIVER_ALGO_VXL_EXPORT threshold
+  : public vital::algo::image_filter
+{
+public:
+  PLUGIN_INFO( "vxl_threshold",
+               "Threshold at image at a given percentile." )
```
Comment (mwoehlke-kitware) [inline arrows/vxl/threshold.h:26]:
```suggestion
               "Threshold at image at a given percentile or value." )
```

**Review by mwoehlke-kitware — COMMENTED** [review]
As a general observation, I wonder if the number of sampling points should be configurable?

**Comment by russelldj** [comment]
> As a general observation, I wonder if the number of sampling points should be configurable?

Perhaps, but I think this can be handled later if it becomes needed and/or is a bottleneck.

File: arrows/vxl/image_statistics.h (hunk 3 above)
Comment (russelldj) [reply arrows/vxl/image_statistics.h:139]: @mwoehlke-kitware here's another subtle example of uniform initialization going sideways. This is supposed to be a one-length vector containing `percentile` but it looks like on g++ 7.5.0 it treats it as a two-length vector.

**Review by mleotta — APPROVED** [review]

**Comment by mleotta** [comment]
Jenkins test this please

**Comment by dstoup** [comment]
Jenkins test this please

File: arrows/vxl/image_statistics.h (hunk 3 above)
Comment (mwoehlke-kitware) [reply arrows/vxl/image_statistics.h:139]: ah, yes, the cursed `vector` ctor strikes again...

BTW, `std::vector< double >{ percentile }` also works.

**Comment by russelldj** [comment]
@mwoehlke-kitware are you comfortable with this landing?

**Review by mwoehlke-kitware — APPROVED** [review]

**Comment by russelldj** [comment]
This looks like it should be able to land now.

**Pull request merged**